File: src/pillar_based_removal/src/pillar_based_removal_node.cpp
#include<pillar_based_removal/pillar_based_removal.h>



template<class T>
void PillarBasedRemoval::declare_param(const std::string &name, 
                                       const T &default_value, 
                                       const std::string &description)
{
  rcl_interfaces::msg::ParameterDescriptor desc;
  desc.description = description;
  this->declare_parameter(name, default_value, desc);
}

void PillarBasedRemoval::set_params()
{
  declare_param<bool>("verbose", false, "If set true, the description of each module will be verbosely printed in the terminal");
  verbose_ = get_parameter("verbose").as_bool();

  declare_param<std::string>("subscription_name", "point_cloud", "This parameter decides the name of the node subscribed");
  subscription_name_ = get_parameter("subscription_name").as_string();
  declare_param<std::string>("publisher_name", "reduced_point_cloud", "This parameter decides the name of the publisher");
  publisher_name_ = get_parameter("publisher_name").as_string();

  declare_param<std::string>("device", "cpu", "This parameter decides which device you are going to use, 1. cpu (default), 2. cuda");
  declare_param<double>("resolution", 0.4, "This parameter decides the side length of the square pillar, the unit is [meter]");
  declare_param<int>("max_num_pillars", 30000, "This parameter decides how many pillars are allowed, set it to a reasonably"
                                            "large number to cover all the pillars.");
  declare_param<std::vector<double>>("lidar_ranges", std::vector<double>{-90, -90, 90, 90}, "This parameter accepts a vector which represents "
                                                    "[xmin, ymin, xmax, ymax] of the lidar, the unit is [meter]");
  declare_param<double>("environment_radius", 1.8, "This parameter decides the radius of the environment to check in "
                                             "the removal stage, the unit is [meter]");
  declare_param<double>("env_min_threshold", 0.45, "This parameter is a threshold for comparing each pillar's lowest point with surrounding "
                                                    "environment's lowest point, the unit is [meter]");
  declare_param<double>("max_min_threshold", 0.45, "This parameter is a threshold for comparing each pillar's highest point with its lowest point, "
                                                    "the unit is [meter]");
  declare_param<std::vector<double>>("rebuild_radiuses", std::vector<double>{1.8, 5.4}, "This parameter decides how large area to rebuild the "
                                                                   "surrounding ground, multiple values mean we want to "
                                                                   "use different radius for different ranges, the unit "
                                                                   "is [meter]");
  declare_param<std::vector<double>>("range_split", std::vector<double>{30}, "This parameter denotes the split of different ranges when using "
                                                         "different radius. It should correspond to the number of radiuses "
                                                         "used. empty for global rebuild, the unit is [meter]");

  device_ = get_parameter("device").as_string();
  device_num_ = device_ == "cpu" ? -1 : 0;
  resolution_ = get_parameter("resolution").as_double();
  max_num_pillars_ = get_parameter("max_num_pillars").as_int();
  lidar_ranges_ = get_parameter("lidar_ranges").as_double_array();
  environment_radius_ = get_parameter("environment_radius").as_double();
  env_min_threshold_ = get_parameter("env_min_threshold").as_double();
  max_min_threshold_ = get_parameter("max_min_threshold").as_double();
  rebuild_radiuses_ = get_parameter("rebuild_radiuses").as_double_array();
  range_split_ = get_parameter("range_split").as_double_array();

  num_received_points_ = 0;
  num_send_points_ = 0;
  
}


void PillarBasedRemoval::msgToTensor(const sensor_msgs::msg::PointCloud2 &received_point_cloud_msg) {
  auto t0 = std::chrono::steady_clock::now();

  pcl::fromROSMsg<PointT>(received_point_cloud_msg, received_point_cloud_);
  num_received_points_ = received_point_cloud_.points.size();
  
  point_cloud_tensor_ = tv::zeros({num_received_points_, 3}, tv::type_v<float>, device_num_);
  auto point_cloud_tensor_ptr = point_cloud_tensor_.data_ptr<float>();

  tv::dispatch<float, float>(point_cloud_tensor_.dtype(), [&](auto I) {
    using T = decltype(I);
    auto point_cloud_tensor_rw = point_cloud_tensor_.tview<T, 2>();
    for(size_t i = 0; i < num_received_points_; i++) {
      point_cloud_tensor_rw((int)i, 0) = received_point_cloud_.points[i].x;
      point_cloud_tensor_rw((int)i, 1) = received_point_cloud_.points[i].y;
      point_cloud_tensor_rw((int)i, 2) = received_point_cloud_.points[i].z;
    }
  });

  auto t1 = std::chrono::steady_clock::now();
  std::chrono::duration<double> time_used_ = std::chrono::duration_cast<std::chrono::duration<double>>(t1 - t0);
  
  if(verbose_)
    RCLCPP_INFO(get_logger(), "The time needed for converting %d points to tensor is %f.",
                              (int) num_received_points_, time_used_.count());
}

void PillarBasedRemoval::tensorToMsg() {
  // Save all the valid indexes
  std::unordered_set<int> valid_voxel_indices;
  auto kept_pillars_ptr = kept_pillars_.data_ptr<uint16_t>();
  for(int i = 0; i < kept_pillars_.dim(0); i++) {
    if(kept_pillars_ptr[i] > 0) {
      valid_voxel_indices.insert(i);
    }
  }

  PointCloudT target_point_cloud = PointCloudT();
  // target_point_cloud.resize(num_send_points_);
  // int idx = 0;
  num_send_points_ = 0;
  auto point_pillar_idx_ptr = point_pillar_idx_.data_ptr<int64_t>();
  for(int i = 0; i < num_received_points_; i++) {
    if(valid_voxel_indices.count((int) point_pillar_idx_ptr[i])) {
      // target_point_cloud[idx] = received_point_cloud_[i];
      target_point_cloud.push_back(received_point_cloud_[i]);
      num_send_points_++;
      // idx++;
    }
  }

  pcl::toROSMsg(target_point_cloud, target_point_cloud_msg_);

  // Publishing the point cloud message.
  if(verbose_)
    RCLCPP_INFO(get_logger(), "Publishing point cloud, the number of sent points is %zu", num_send_points_);
  pointcloud_publisher_->publish(target_point_cloud_msg_);
}


PillarBasedRemoval::PillarBasedRemoval() : Node("pillar_based_removal_node") {
  set_params();
  pointcloud_subscriber_ = create_subscription<sensor_msgs::msg::PointCloud2>(
    subscription_name_, rclcpp::QoS(rclcpp::KeepLast(1)).best_effort().durability_volatile(),
             std::bind(&PillarBasedRemoval::callback, this, std::placeholders::_1)
  );
  pointcloud_publisher_ = create_publisher<sensor_msgs::msg::PointCloud2>(
    publisher_name_, rclcpp::QoS(rclcpp::KeepLast(1)).best_effort().durability_volatile()
  );
}

PillarBasedRemoval::~PillarBasedRemoval() {}


void PillarBasedRemoval::callback(const sensor_msgs::msg::PointCloud2 &received_point_cloud_msg) {
  auto t0 = std::chrono::steady_clock::now();

  msgToTensor(received_point_cloud_msg);
  pillarize();
  removal_stage();
  rebuild_stage();
  tensorToMsg();

  auto t1 = std::chrono::steady_clock::now();
  std::chrono::duration<double> time_used_ = std::chrono::duration_cast<std::chrono::duration<double>>(t1 - t0);
  RCLCPP_INFO(get_logger(), "Overall time needed for the algorithm is %f sec", time_used_.count());
}



int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PillarBasedRemoval>());
  rclcpp::shutdown();

  return 0;
}

File: src/pillar_based_removal_py/pillar_based_removal_py/pillar_based_removal_node.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import ParameterDescriptor
import array

import sensor_msgs_py.point_cloud2 as pc2

import torch
import numpy as np
from sensor_msgs.msg import PointCloud2
from .pillar_based_removal_algo import *
import time

class PillarBasedRemoval(Node):
    def __init__(self):
        super().__init__("pillar_based_removal_node")
        self.set_params()

        self.subscriber = self.create_subscription(
            PointCloud2,
            self.subscription_name_,
            self.callback,
            1
        )

        self.publisher = self.create_publisher(
            PointCloud2,
            self.publisher_name_,
            1
        )

    
    def declare_param(self, param_name: str, default, description: str):
        descriptor = ParameterDescriptor(description=description)
        self.declare_parameter(param_name, default, descriptor)

    def set_params(self):
        self.declare_param("verbose", True, "If set true, the description of each module will be verbosely printed in the terminal")
        self.verbose_ = self.get_parameter("verbose").get_parameter_value().bool_value

        self.declare_param("subscription_name", "kitti/velo", "This parameter decides the name of the node subscribed")
        self.subscription_name_ = self.get_parameter("subscription_name").get_parameter_value().string_value
        self.declare_param("publisher_name", "reduced_point_cloud", "This parameter decides the name of the publisher")
        self.publisher_name_ = self.get_parameter("publisher_name").get_parameter_value().string_value
        self.declare_param("point_fields", ["x", "y", "z", "intensity"], "Please specify the point cloud's field names and ensure the \
                                                                            first three indicate xyz values")
        self.declare_param("enable_rebuild", True, "If set to False, then the rebuild stage will not take effect")
        self.enable_rebuild_ = self.get_parameter("enable_rebuild").get_parameter_value().bool_value
        self.point_fileds_ = self.get_parameter("point_fields").get_parameter_value().string_array_value
        self.declare_param("device", "cuda", "This parameter decides which device you are going to use, 1. cpu (default), 2. cuda")
        self.declare_param("resolution", 0.4, "This parameter decides the side length of the square pillar, the unit is [meter]")
        self.declare_param("max_num_pillars", 50000, "This parameter decides how many pillars are allowed, set it to a reasonably"
                                                    "large number to cover all the pillars.")
        self.declare_param("lidar_ranges", [-90., -90., 90., 90.], "This parameter accepts a vector which represents "
                                                            "[xmin, ymin, xmax, ymax] of the lidar, the unit is [meter]")
        self.declare_param("environment_radius", 1.8, "This parameter decides the radius of the environment to check in "
                                                    "the removal stage, the unit is [meter]")
        self.declare_param("env_min_threshold", 0.45, "This parameter is a threshold for comparing each pillar's lowest point with surrounding "
                                                            "environment's lowest point, the unit is [meter]")
        self.declare_param("max_min_threshold", 0.45, "This parameter is a threshold for comparing each pillar's highest point with its lowest point, "
                                                            "the unit is [meter]")
        self.declare_param("rebuild_radiuses", [1.8, 5.4], "This parameter decides how large area to rebuild the "
                                                                        "surrounding ground, multiple values mean we want to "
                                                                        "use different radius for different ranges, the unit "
                                                                        "is [meter]")
        self.declare_param("range_split", [30.], "This parameter denotes the split of different ranges when using "
                                                                "different radius. It should correspond to the number of radiuses "
                                                                "used. empty for global rebuild, the unit is [meter]")

        self.device_ = self.get_parameter("device").get_parameter_value().string_value
        self.device_num_ = -1 if self.device_ == "cpu" else 0
        self.resolution_ = self.get_parameter("resolution").get_parameter_value().double_value
        self.max_num_pillars_ = self.get_parameter("max_num_pillars").get_parameter_value().integer_value
        self.lidar_ranges_ = self.get_parameter("lidar_ranges").get_parameter_value().double_array_value
        self.environment_radius_ = self.get_parameter("environment_radius").get_parameter_value().double_value
        self.env_min_threshold_ = self.get_parameter("env_min_threshold").get_parameter_value().double_value
        self.max_min_threshold_ = self.get_parameter("max_min_threshold").get_parameter_value().double_value
        self.rebuild_radiuses_ = self.get_parameter("rebuild_radiuses").get_parameter_value().double_array_value
        self.range_split_ = self.get_parameter("range_split").get_parameter_value().double_array_value

        self.num_received_points_ = 0
        self.num_send_points_ = 0

    def msgToTensor(self, msg: PointCloud2):
        start = time.time()

        self.received_point_cloud_msg_ = msg
        
        data = pc2.read_points(msg, None, True)
        num_points = len(data)
        self.resolved_data = data
        t1 = time.time()
        
        self.point_cloud_tensor_ = torch.zeros((num_points, 3), device=torch.device(self.device_))

        x_tensor = torch.tensor(np.copy(data[self.point_fileds_[0]]), dtype=torch.float32, device=torch.device(self.device_))
        y_tensor = torch.tensor(np.copy(data[self.point_fileds_[1]]), dtype=torch.float32, device=torch.device(self.device_))
        z_tensor = torch.tensor(np.copy(data[self.point_fileds_[2]]), dtype=torch.float32, device=torch.device(self.device_))

        # Stack these tensors along a new dimension (dim=1) to form a matrix
        self.point_cloud_tensor_ = torch.stack([x_tensor, y_tensor, z_tensor], dim=1)

        self.point_cloud_tensor_ = self.point_cloud_tensor_.contiguous()
        self.num_received_points_ = self.point_cloud_tensor_.shape[0]
        

        end = time.time()
        duration = end - start
        if self.verbose_:
            self.get_logger().info("Time needed for resolving received points is %f, %f, %f" % (duration, t1-start, end-t1))
    
    def tensorToMsg(self):
        start = time.time()

        self.kept_pillars_ = (self.kept_pillars_ > 0).squeeze()
        pillar_indices = torch.cumsum(torch.ones((len(self.kept_pillars_), ), device=self.kept_pillars_.device), dim=0) - 1
        self.kept_pillars_ = pillar_indices[self.kept_pillars_]

        valid_points_indices = torch.isin(self.point_pillar_idx_, self.kept_pillars_).cpu().numpy()
        resolved_data = self.resolved_data[valid_points_indices]
        self.num_send_points_ = np.sum(valid_points_indices)

        memory_view = memoryview(resolved_data)
        casted = memory_view.cast('B')
        array_array = array.array('B')
        array_array.frombytes(casted)

        sending_point_cloud_msg = self.received_point_cloud_msg_
        sending_point_cloud_msg.data = array_array
        sending_point_cloud_msg.width = len(resolved_data)
        sending_point_cloud_msg.height = 1
        sending_point_cloud_msg.is_dense = False
        sending_point_cloud_msg.header.stamp = self.get_clock().now().to_msg()
        
        self.publisher.publish(sending_point_cloud_msg)

        end = time.time()
        duration = end - start
        if self.verbose_:
            self.get_logger().info("Time needed for resolving target points is %f" % duration)

    def save_points(self, data, path):
        x = data[self.point_fileds_[0]]
        y = data[self.point_fileds_[1]]
        z = data[self.point_fileds_[2]]
        points = np.stack((x, y, z), axis = 1)

        np.save(path, points)
        self.get_logger().info("Points saved to %s" % path)

    def callback(self, msg: PointCloud2):
        start = time.time()

        self.msgToTensor(msg)
        self.pillarize()
        self.removal_stage()
        self.rebuild_stage()
        self.tensorToMsg()

        reduced_num = self.num_received_points_ - self.num_send_points_
        reduced_percentage = reduced_num / self.num_received_points_

        end = time.time()
        duration = end - start
        self.get_logger().info("Overall time needed is {:.5f} seconds, reduced {:.2%}".format(duration, reduced_percentage))


PillarBasedRemoval.pillarize = pillarize
PillarBasedRemoval.removal_stage = removal_stage
PillarBasedRemoval.rebuild_stage = rebuild_stage

def main(args=None):
    rclpy.init(args=args)

    pillar_based_removal_node = PillarBasedRemoval()

    rclpy.spin(pillar_based_removal_node)

    pillar_based_removal_node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()